c++/src/kj/async.h

```diff
@@ -202,6 +202,11 @@ class [[nodiscard]] Promise: protected _::PromiseBase {
   template <typename Func>
   PromiseForResult<Func, T> then(Func&& func) KJ_WARN_UNUSED_RESULT;
 
+  // Specialized variants for Promise<void> continuations. This results in substantial code size
+  // improvements.
+  Promise<void> thenV(kj::Function<void()>&& func) requires std::is_void_v<T> KJ_WARN_UNUSED_RESULT;
```

Comment:
kj::Function introduces significant overhead -- an extra heap allocation on construction and virtual function dispatch when called. Have you measured the performance impact?

Reply:
That is unfortunate. I agree that allocation overhead is not desirable here and assumed kj::Function was just a thin wrapper around C-style function pointers. C++ doesn't allow using function pointers for non-capturing lambdas and almost all of the impact comes from capturing lambdas, so I guess eliminating Func as a parameter to SimpleTransformPromiseNode just won't be feasible here.

Reply:
Yeah function pointers fundamentally can't work with captures, because where do the captures go? And the captures are also why we need either templates or dynamic allocation + virtual dispatch.